design/dynamic-provision-filesystem.md

```diff
@@ -0,0 +1,137 @@
+Also Dynamic Provision Filesystem is a concept that has already been done in https://github.com/kubernetes-incubator/external-storage/tree/master/ceph/cephfs. So Rook can adopt a similar approach.
+
+## Current Experience
+
+To consume a filesystem, the experience is the following:
+
+```yaml
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: mysql
+spec:
+  strategy:
+    type: Recreate
+  template:
+    spec:
+      containers:
+      - image: mysql:5.6
+        name: mysql
+        env:
+        - name: MYSQL_ROOT_PASSWORD
+          value: changeme
+        ports:
+        - containerPort: 3306
+          name: mysql
+        volumeMounts:
+        - name: mysql-persistent-storage
+          mountPath: /var/lib/mysql
+      volumes:
+      - name: mysql-persistent-storage
+        cephfs:
+          monitors:
+          - monitor1
+          - monitor2
+          - monitor3
+          user: admin
+          secretRef:
+            name: rook-admin
+```
+
+Users will have to come up with these values and ensure every parameter is provided correctly.
+
+## Experience with Dynamic Provisioned Filesystem
+
+To create a filesystem, you just create a PVC object. This is consistent with all other storage provisioning in Kubernetes.
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: myfsData
+spec:
+  storageClassName: rook-filesystem-simple
+  path: /myData # Will use root path, "/", if not provided
+  accessModes:
+  - ReadWriteMany
+```
+
+To consume it, the pod manifest is shown as follows:
+
+```yaml
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: mysql
+spec:
+  strategy:
+    type: Recreate
+  template:
+    spec:
+      containers:
+      - image: mysql:5.6
+        name: mysql
+        env:
+        - name: MYSQL_ROOT_PASSWORD
+          value: changeme
+        ports:
+        - containerPort: 3306
+          name: mysql
+        volumeMounts:
+        - name: mysql-persistent-storage
+          mountPath: /var/lib/mysql
+      volumes:
+      - name: mysql-persistent-storage
+        persistentVolumeClaim:
+          claimName: myfsData
+```
+
+Note that the consuming pod manifest looks the same whether it is mounting a filesystem or a block device.
+
+## StorageClass Example
+
+Notice that there was a reference to a StorageClass called `rook-filesystem-simple` in the filesystem PVC example I previously showed. Dynamic provisioned storage refers to a StorageClass object for details and configuration about how the storage should be provisioned.
+The storage class is setup by the administrator and can look as follows for filesystem:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: rook-filesystem-simple
+provisioner: rook.io/filesystem
+parameters:
+  fsName: myFS # Name of the filesystem to use.
+```
+
+The referenced filesystem, `myFS`, would have to be also created by the admin using a [Filesystem CRD](/Documentation/pool-crd.md).
```

Comment:
so the admin manually creates the file system using a CRD and creates the storage class too, right?

Reply:
correct. The users will then reference the storageclass that references their desired filesystem.

Reply:
it would be helpful to show the sequence of what is needed, although i suppose it's a documentation item rather than a design doc question. For example, first create the CRD, then the storage class, then consume it with the pvc.